contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,223 @@
+	Exclusive        bool
+	NoWait           bool
+}
+
+type Adapter struct {
+	Brokers        string
+	Topic          string
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+	SinkURI        string
+	client         client.Client
+}
+
+func (a *Adapter) initClient() error {
+	if a.client == nil {
+		var err error
+		if a.client, err = kncloudevents.NewDefaultClient(a.SinkURI); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Start(ctx context.Context, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		panic(err)
+	}
+	defer ch.Close()
+
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err = ch.ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		exchangeConfig.Durable,
+		exchangeConfig.AutoDeleted,
+		exchangeConfig.Internal,
+		exchangeConfig.NoWait,
+		nil,)
+	if err != nil {
+		panic(err)
+	}
+
+	queue, err := ch.QueueDeclare(
+		a.QueueConfig.Name,
+		a.QueueConfig.Durable,
+		a.QueueConfig.DeleteWhenUnused,
+		a.QueueConfig.Exclusive,
+		a.QueueConfig.NoWait,
+		nil,)
+	if err != nil {
+		panic(err)
+	}
+
+	routingKeys := strings.Split(a.QueueConfig.RoutingKey, ",")
+
+	for _, routingKey := range routingKeys {
+		err = ch.QueueBind(
+			queue.Name,
+			routingKey,
+			exchangeConfig.Name,
+			a.QueueConfig.NoWait,
+			nil,)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if err := a.initClient(); err != nil {
+		logger.Error("Failed to create cloudevent client", zap.Error(err))
+		return err
+	}
+
+	return a.pollForMessages(ctx, ch, &queue, stopCh)
+}
+
+func (a *Adapter) pollForMessages(ctx context.Context, channel *amqp.Channel,
+	queue *amqp.Queue, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	msgs, err := channel.Consume(
+		queue.Name,
+		"",
+		true,
+		a.QueueConfig.Exclusive,
+		false,
+		a.QueueConfig.NoWait,
+		nil,)
+
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		select {
+		case msg, ok := <-msgs:
+			if ok {
+				logger.Info("Received: ", zap.Any("value", string(msg.Body)))
+
+				if err := a.postMessage(ctx, &msg); err != nil {
+					logger.Info("Error posting message: ", zap.Error(err))
+				}
+			} else {
+				return nil
+			}
+		case <-stopCh:
+			logger.Info("Shutting down...")
+			return nil
+		}
+	}
+}
+
+func (a *Adapter) postMessage(ctx context.Context, msg *amqp.Delivery) error {
+	logger := logging.FromContext(ctx)
+
+	extensions := map[string]interface{}{
+		"key": string(msg.MessageId),
+	}
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			SpecVersion: cloudevents.CloudEventsVersionV02,
+			Type:        eventType,
+			ID:          msg.MessageId,
+			Time:        &types.Timestamp{msg.Timestamp},
+			Source:      *types.ParseURLRef(a.Brokers),
```

Comment:
I think only Brokers information is not enough for a `Source` information. At least, exchangeConfig.Name or queue.Name shall be added.